### DISCUSSION TITLE
Vercel mutates url slug in an unexpected way.

### DISCUSSION THREAD

**Discussion by FarahZaqout** [vercel/community#133] [Help] [answered]
Happy holidays everyone.

I work on a Nextjs project deployed on cloudfront, while testing environments are deployed on heroku. We are planning to move to Vercel and whenever we try to test a vercel deployment for the same code, the app returns 404 for existing pages in certain scenarios, despite the code being the same while it works fine on other platform.

## On Vercel:
* A slug with a single word, for example `/doctors/urologist` will work just fine.
* A slug with multiple words; for example, `/doctors/general+practitioners` will work fine if we go to it via a navbar link. However, it will throw a `404` error if we refresh or copy the url directly into the browser.
* A slug with multiple words separated by `-`, `/doctors/pediatrician-aetna` will work fine.

We took a look at requests coming to our backend servers, and found the urls that are separated by `+` are encoded with `%20` in place of the `+` separator, which is the encoding for `space` rather than `+`.  It looks like this `general%20practitioners`, which corresponds with `general practitioners`, this we receive a 404 error.

## To reproduce the behavior:
* Go to :https://vercel.heliumdoc.com/
* From the Navbar, click the specialties dropdown menu, and select a specialization with more than one word. Once it loads, refresh and it will give a 404 page.
* Select a single word specialty and after the page loads refresh. It will work just fine.

None of this happens when we are testing deployments on Heroku or Cloudfront. Help would be highly appreciated, I am hard stuck on this. 

Thanks.

**Comment by Pieparker** [comment, marked as answer]
Thanks for your patience and good find! We don't honour the "+" character and, rather, serve it encoded with `%2b` e.g. 
> https://vercel.heliumdoc.com/doctors/general%2bpractitioner

**Reply by akornato** [reply]
@Pieparker why don't you honor the plus character? This is sometimes breaking our app, where one of query params has an ID which may include plus character. Also you don't replace it encoded with `%2b` but with `%20`, which makes no sense.

**Reply by maerzhase** [reply]
@Pieparker i would also follow up on this: Could you at least provide some documentation on which characters you honour and which not? this is a pitfall for anyone using vercel + slugs because you basically don't get this behaviour locally but only after deployment.

**Reply by amyegan** [reply]
The `+` character is a [reserved character](https://en.wikipedia.org/wiki/Percent-encoding#Reserved_characters) with special meaning in some contexts. Depending on how your app is set up, it may be automatically encoded as `%2B`. I suspect something  might be interfering for the original poster to get `%20` instead. Possibly a misconfigured Serverless Function or Middleware altering the URL. Difficult to say without seeing it in action or digging into the code.

I just tried deploying a path with a `+` character to my test site and it is working for me. But my test site is very simple and likely quite different from your project. Depending on your tech stack and what you want to accomplish, there could be ways to make it work. If you [create a new discussion](https://github.com/vercel/community/discussions/new) with details about your situation, then we can work together to find a solution.

**Reply by maerzhase** [reply]
@amyegan 
I created a minimal reproduction example, which shows the pitfall:
Locally the plus can be encoded. On vercel the plus is simply removed. 

# local environment
<img width="497" alt="Screenshot 2022-09-21 at 13 04 25" src="https://user-images.githubusercontent.com/1128485/191488655-0266beee-003d-47aa-9a16-006caea4dae8.png">

# vercel
<img width="641" alt="Screenshot 2022-09-21 at 13 04 31" src="https://user-images.githubusercontent.com/1128485/191488683-a99c498c-0a5f-467e-8455-c8227c72ac46.png">

Deployment is here: https://nextjs-vercel-slug-example.vercel.app/slug+slug
Repository is here: https://github.com/maerzhase/nextjs-vercel-slug-example/blob/main/pages/%5Bslug%5D/index.tsx

**Reply by amyegan** [reply]
Perfect example! Thank you for sharing this. I'm going to ask the team and see what info I can dig up.

**Reply by maerzhase** [reply]
@amyegan any news ? 🙈

**Reply by amyegan** [reply]
Sorry for the slow updates. 😞 No news yet, but the team is investigating since this does seem like a bug

**Reply by amyegan** [reply]
No solution yet, but I wanted to let you know that the team has not forgotten about this. Thank you all for your patience while we investigate ways to change this behavior without breaking things for people who have come to rely on this quirk.

**Reply by maerzhase** [reply]
thanks for all the effort @amyegan !

**Reply by amyegan** [reply]
After digging into the details and history of this behavior, this is being considered a long-standing behavior with no immediate plans to change. A strategy will need to be devised to change this behavior without breaking existing projects that rely on it. The team will revisit this in future, but we don't have a concrete ETA on a change at present.

**Reply by jakepeyser** [reply]
I appreciate you all looking into this behavior, it just came up while migrating my app to Vercel as we rely on the `+` character to separate some subpaths in our URLs.

@amyegan I'd like to voice support for at least treating these characters as `+` literals and encoding them as `%2B` rather than the space character. I've had to add special decoding logic to my app route components to account for this in the meantime.

**Comment by ghost** [comment]
Is this documented anywhere on Vercel? I encountered it today and as @maerzhase pointed out, it works differently locally vs deployed to Vercel which made it difficult to debug. 

We want to ensure the correct URL is being used to access the dynamic route by forcing it to match the page title, but the existing URL pattern includes `+` which is incorrectly encoded as `%20` by Vercel and causes a redirect loop because it is always a mismatch.
